usb: udc: numaker: support numaker m46x/m55m1x hsusbd

ccli8 opened the pull request:
This adds support for nuvoton numaker high-speed USB 2.0 device (`hsusbd`) and makes some fixes, including:
1. Support m46x/m55m1x `hsusbd`. This re-organizes `udc_numaker.c` to support both `usbd` and `hsusbd` in single file, but they cannot enable simultaneously.
2. Recycle dangling `net_buf` to recover from incomplete control transfer
3. Other fixes on `usbd`

ccli8 committed:
drivers: usb: udc: numaker: usbd: refine MXPLD read

For USBD, this refines MXPLD read to avoid unwanted or reserved bits.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: numaker: report instead of assert in control transfer

On Control Out failure, this changes to report error message instead of
assert failure because USB bus error is allowed to happen.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: numaker: recover from incomplete control transfer

Previous control transfer can be incomplete and then causes not only
net_buf leak but also logic error. This recycles dangling net_buf for
new clean control transfer.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: numaker: usbd: enable usb wake-up early

Same as VBUS detect, this enables USB wake-up early so that device
can also be woken up by VBUS.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: numaker: support NuMaker M55M1X HSUSBD

Add support for Nuvoton NuMaker M55M1X high-speed USB 2.0 device
controller. Compared to M46X HSUSBD, M55M1X HSUSBD introduces some
differences:
- DMA must handle cache coherency because net_buf can be cache-able
- USB suspend interrupt becomes continuous. Aavoid being locked by
  this interrupt and forward this message just once
- New register bit HSUSBD_OPER_HISHSEN_Msk, which controls to enable
  USB handshake

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: numaker: support NuMaker M46X HSUSBD

Add support for Nuvoton NuMaker M46X high-speed USB 2.0 device
controller.

The code is re-organized to implement both usbd and hsusbd in single
source file. Multiple instances of either usbd or hsusbd are supported,
but usbd and hsusbd cannot support simultaneously. This limitation is
for easy implementation with just single source file, assuming that real
application just needs one usb device type.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: numaker: usbd: refine on ATTR undefined bits

Refine code on ATTR undefined bits:
- Add USBD_ATTR_PWRDN_Msk if it is not defined
- Following BSP USBD driver, add note on BIT(6) for hidden

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -731,62 +1284,237 @@ static void numaker_usbd_ep_config_major(struct numaker_usbd_ep *ep_cur,
 	/* Disable EP */
 	ep_cur->ep_hw_cfg = (ep_cur->ep_hw_cfg & ~USBD_CFG_STATE_Msk) | USBD_CFG_EPMODE_DISABLE;
 	ep_base->CFG = ep_cur->ep_hw_cfg;
+#elif DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_hsusbd)
```
`if` will add one more nesting level and influence code layout. It seems `#if` is more common for such exclusive code, like: https://github.com/zephyrproject-rtos/zephyr/blob/73ea9419c354526edf175607c5f0713f892a5196/drivers/usb/udc/udc_stm32.c#L1048-L1050

jfischer-no commented on drivers/usb/udc/Kconfig.numaker:
I think your commit message can be improved. Please do not use a numbered list in a commit message. Something like

```
drivers: udc: numaker: support NuMaker M46X HSUSBD

Add support for Nuvoton NuMaker M46X high-speed USB 2.0 device
controller.
...

```
should be fine here.

jfischer-no commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -31,36 +29,77 @@ LOG_MODULE_REGISTER(udc_numaker, CONFIG_UDC_DRIVER_LOG_LEVEL);
  *        to generate necessary 48MHz.
  */
 
+BUILD_ASSERT(!(DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_usbd) &&
+	       DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_hsusbd)),
+	     "NOT SUPPORT USBD/HSUSBD simultaneously");
```
Why not?

jfischer-no commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -31,36 +29,77 @@ LOG_MODULE_REGISTER(udc_numaker, CONFIG_UDC_DRIVER_LOG_LEVEL);
+
 /* For bus reset, keep 'SE0' (USB spec: SE0 >= 2.5 ms) */
-#define NUMAKER_USBD_BUS_RESET_DRV_SE0_US 3000
+#define NUMAKER_XUSBD_BUS_RESET_DRV_SE0_US 3000
```
These renaming changes, NUMAKER_USBD->NUMAKER_XUSBD numaker_usbd->numaker_xusbd, does not really make sense as the driver still supports USB 1.1 controller. It only makes the review harder. Therefore please drop these changes.

jfischer-no commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -635,61 +1064,183 @@ static void numaker_usbd_ep_copy_to_user(struct numaker_usbd_ep *ep_cur, uint8_t
 
 	*size_p = MIN(*size_p, data_rmn);
 
+#if DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_usbd)
 	bytecpy(usrbuf, (uint8_t *)dmabuf_addr, *size_p);
+#elif DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_hsusbd)
+	if (ep_cur->ep_hw_idx == CEP) {
+		uint8_t *usrbuf_pos = usrbuf;
+		uint32_t rmn = *size_p;
+
+		while (rmn && !(base->CEPINTSTS & HSUSBD_CEPINTSTS_BUFEMPTYIF_Msk)) {
+			*usrbuf_pos++ = base->CEPDAT_BYTE;
+			rmn--;
+		}
+
+		*size_p -= rmn;
+	} else {
+#if defined(CONFIG_UDC_NUMAKER_DMA)
+		int err = numaker_xusbd_ep_xfer_user_dma(ep_cur, usrbuf, size_p);
+
+		if (err < 0) {
+			return err;
+		}
+#else
+		uint8_t *usrbuf_pos = usrbuf;
+		uint32_t rmn = *size_p;
+
+		while (rmn && !(ep_base->EPINTSTS & HSUSBD_EPINTSTS_BUFEMPTYIF_Msk)) {
+			*usrbuf_pos++ = ep_base->EPDAT_BYTE;
+			rmn--;
+		}
+
+		*size_p -= rmn;
+#endif
+	}
+#endif
+
```
This part and similar ones below do not look good. Please rework it/move it to specific functions.

ccli8 replied:
They are moved to separate functions for clear.

jfischer-no commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -567,66 +814,248 @@ static void numaker_usbd_ep_th(const struct device *dev, uint32_t ep_hw_idx)
 	__ASSERT_NO_MSG(size_p);
 	__ASSERT_NO_MSG(ep_cur->dmabuf_valid);
 
+#if DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_usbd)
```
Something like `DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_usbd)` should not be used in the code, the driver has to support multiple instances of the controller on the SoC.

jfischer-no commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -1016,6 +1050,15 @@ static int numaker_xusbd_ep_xfer_user_dma(struct numaker_xusbd_ep *ep_cur, uint8
 	base->DMACTL |= USB_EP_GET_IDX(ep_cur->addr) << HSUSBD_DMACTL_EPNUM_Pos;
 
+	/* Cache coherency */
+#if !defined(CONFIG_UDC_BUF_FORCE_NOCACHE)
```
Why would be the option UDC_BUF_FORCE_NOCACHE enabled?

ccli8 replied:
Removed. `sys_cache_data` functions can cover no-cache case.

jfischer-no commented on drivers/usb/udc/udc_numaker.c:
>This supports nuvoton numaker m55m1x high-speed USB 2.0 device
controller (hsusbd), including:
>1. Support DMA with cache coherency
>2. Handle USB suspend interrupt becoming continuous
>3. Handle new HSUSBD_OPER_HISHSEN_Msk

Looks like this commit should be spited in a few specific changes.

jfischer-no reviewed (dismissed):
There is an unacceptable amount of 
```
#if DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_usbd)
#elif DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_hsusbd)
#endif
```
you have to find another way to handle it.

ccli8 committed:
drivers: usb: udc: numaker: support HSUSBD DMA

Add DMA support for Nuvoton NuMaker M46X high-speed USB 2.0 device
controller.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: numaker: support USBD/HSUSBD simultaneously

Re-organize to add support for USBD/HSUSBD which can be enabled
at the same time

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 commented:
> There is an unacceptable amount of `#if DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_usbd)`

> How I understood it is limitation in the driver introduced by you, right? I think there are a few ways to allow both controllers to be used. Is there a common hardware register that identifies the controller type?

Now the driver can support USBD/HSUSBD simultaneously, and all conditional code `#if DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_usbd)` become of no use and are removed. This support is by passing controller type to udc device define `DT_INST_FOREACH_STATUS_OKAY(UDC_NUMAKER_DEVICE_DEFINE)` for USBD and HSUSBD. Then all driver code can run-time distinguish between USBD and HSUSBD.

cyliangtw reviewed (approved):
LGTM
